File: src/models/cashflow.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, Index, Numeric
from sqlalchemy.orm import Mapped, mapped_column

from .base_model import BaseDBModel


class CashflowForecast(BaseDBModel):
    """Model for storing cashflow forecasts

    This is a pure data storage model with no business logic.
    All calculation methods have been moved to the service layer.
    """

    __tablename__ = "cashflow_forecasts"

    id: Mapped[int] = mapped_column(primary_key=True)
    forecast_date: Mapped[datetime] = mapped_column(
        DateTime(), nullable=False, doc="UTC timestamp of the forecast (naive UTC)"
    )
    total_bills: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    total_income: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    balance: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    forecast: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    min_14_day: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    min_30_day: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    min_60_day: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    min_90_day: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    daily_deficit: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    yearly_deficit: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    required_income: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    hourly_rate_40: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    hourly_rate_30: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    hourly_rate_20: Mapped[Decimal] = mapped_column(Numeric(12, 4))
    # Create indexes for efficient lookups
    __table_args__ = (Index("idx_cashflow_forecast_date", "forecast_date"),)

    def __repr__(self) -> str:
        return f"<CashflowForecast {self.forecast_date} balance={self.balance}>"

File: frontend/src/types/cashflow.ts
export interface CashflowForecast {
  date: Date;
  total_bills: number;
  total_income: number;
  balance: number;
  forecast: number;
  min_14_day: number;
  min_30_day: number;
  min_60_day: number;
  min_90_day: number;
}

// For API responses
export interface CashflowForecastResponse {
  date: string;
  total_bills: string | number;
  total_income: string | number;
  balance: string | number;
  forecast: string | number;
  min_14_day: string | number;
  min_30_day: string | number;
  min_60_day: string | number;
  min_90_day: string | number;
}

// Transform API response to frontend model
export const transformCashflowForecast = (data: CashflowForecastResponse): CashflowForecast => {
  return {
    date: new Date(data.date),
    total_bills: typeof data.total_bills === 'string' ? parseFloat(data.total_bills) : data.total_bills,
    total_income: typeof data.total_income === 'string' ? parseFloat(data.total_income) : data.total_income,
    balance: typeof data.balance === 'string' ? parseFloat(data.balance) : data.balance,
    forecast: typeof data.forecast === 'string' ? parseFloat(data.forecast) : data.forecast,
    min_14_day: typeof data.min_14_day === 'string' ? parseFloat(data.min_14_day) : data.min_14_day,
    min_30_day: typeof data.min_30_day === 'string' ? parseFloat(data.min_30_day) : data.min_30_day,
    min_60_day: typeof data.min_60_day === 'string' ? parseFloat(data.min_60_day) : data.min_60_day,
    min_90_day: typeof data.min_90_day === 'string' ? parseFloat(data.min_90_day) : data.min_90_day,
  };
};
